### DISCUSSION TITLE
tinyusb+RP2040+FreeRTOS

### DISCUSSION THREAD

**Discussion by TravisRo** [hathach/tinyusb#1492] [General]
When using tinyusb with FreeRTOS SMP on the RP2040 with the picosdk, I have to leave CFG_TUSB_OS defined as OPT_OS_PICO to get it to work.  It seems like since It's running via a FreeRTOS task, I should be using OPT_OS_FREERTOS. Can someone elaborate on why this is and what CFG_TUSB_OS should really be set to? 

If I use OPT_OS_FREERTOS, the device runs and it properly detected but nothing can be sent or received.  I am using the cdc_dual_ports example for testing and the tinyusb source included with the picosdk.

**Comment by TravisRo** [comment]
After upgrading the tinyusb submodule in the picosdk from 0.12 to 0.13, it now works with the OPT_OS_FREERTOS option set which I believe is the correct way to set it up but It would be nice to have a confirmation on that.

Thanks for the great usb library!  This is the best thing going for USB firmware.

**Reply by hathach** [reply]
yeah, `OPT_OS_FREERTOS` should be selected when using with freeRTOS.

**Reply by Filip-Declercq** [reply]
Hi TravisRo. I'ld like to start a project with the same combo, but prefer to start from a example, but didn't find one yet. The example for FreeRTOS that is in the TinyUSB repo doesn't support the RP2040/PICO board. How did you setup your project? Where did you start from?

**Reply by TravisRo** [reply]
> Hi TravisRo. I'ld like to start a project with the same combo, but prefer to start from a example, but didn't find one yet. The example for FreeRTOS that is in the TinyUSB repo doesn't support the RP2040/PICO board. How did you setup your project? Where did you start from?

There is no example that I know of that is exactly what you want. You need to start by cloning the freertos SMP branch then use a combination of the freerots SMP examples and the tinyusb examples.  You'll want to update tinyusb in the picosdk to 0.13 and modify the target_compile_definitions in tinyusb/hw/bsp/rp2040/family.cmake such that CFG_TUSB_OS=OPT_OS_FREERTOS

That's it from a 50k foot view.

**Reply by Filip-Declercq** [reply]
Hi, thank you for your reply. I'll give that path a try!
